### PR TITLE
Incorporate domain knowledge modifications

### PR DESCRIPTION
**Pull request by ibrahimsharaf**
Domain-aware training for sentiment analysis, mainly inspired from (https://www.ijcai.org/proceedings/2019/0681.pdf).

### CODE REVIEW

**Comment by Mostafa-Samir** [comment]
I prefer to have this gone to a new branch for domain training based form `train_split` instead of overwriting `train_split`

File: drnn.py [hunk 1]
```diff
@@ -117,9 +121,54 @@ def logits(self):
...
+                tf.nn.rnn_cell.GRUCell(
+                    self.hidden_size,
+                    kernel_initializer=tf.contrib.layers.xavier_initializer(False),
+                    bias_initializer=tf.zeros_initializer(),
+                ),
+                output_keep_prob=self.hidden_dropout_keep_prob)
+
+            outputs, _= tf.nn.bidirectional_dynamic_rnn(
+                cell_fw=biGRU_layer,
+                cell_bw=biGRU_layer,
```
Comment (Mostafa-Samir) [inline drnn.py]: Use different cells for the forward and backward directions

File: drnn.py [hunk 2]
```diff
@@ -171,59 +220,63 @@ def logits(self):
 
                 logging.info('state_output_batch_norm: {}'.format(state_output_batch_norm.shape))
 
-                W1 = tf.get_variable("W1", shape=[self.hidden_size, 200],
+                W1 = tf.get_variable("W1", shape=[self.hidden_size, 300],
                                      initializer=tf.initializers.he_normal())
-                b1 = tf.Variable(tf.constant(0, shape=[200], dtype=tf.float32), name="b1")
-                mlp_output = tf.nn.relu(
+                b1 = tf.Variable(tf.constant(0, shape=[300], dtype=tf.float32), name="b1")
```
Comment (Mostafa-Samir) [inline drnn.py]: Use `get_variable` instead of `tf.Variable`

File: drnn.py [hunk 3]
```diff
@@ -171,59 +220,63 @@ def logits(self):
...
+                kernel_initializer=tf.initializers.he_normal()
+            )
+            B_hat = tf.layers.dense(
+                B,
+                self.sequence_length,
+                activation=None,
+                kernel_initializer=tf.initializers.he_normal()
+            )
+            attention_scores = tf.nn.softmax(B_hat)
+            sentiment_attended = tf.reduce_sum(tf.matmul(attention_scores, sentiment_drnn_features), axis=1)
```
Comment (Mostafa-Samir) [inline drnn.py]: This is not the right way to implement the equations:
image
the multiplication between the attention weights and the transformed sentiment DRNN features is an element wise multiplication, not a matrix multiplication. It should be written like that:
```python
setiment_attended = tf.reduce_sum(
    tf.expand_dims(attention_scores, axis=2) * sentiment_drnn_features,
    axis = 1
)
```

File: drnn.py [hunk 4]
```diff
@@ -171,59 +220,63 @@ def logits(self):
...
 
     @define_scope
     def optimize(self):
-        loss = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=self.input_y, logits=self.logits)
-        self.loss = tf.reduce_mean(loss)
+        sentiment_loss = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=self.input_y, logits=self.logits[0])
+        domains_loss = tf.nn.sparse_softmax_cross_entropy_with_logits(labels=self.domains_y, logits=self.logits[1])
+        self.sentiment_loss = tf.reduce_mean(sentiment_loss)
+        self.domains_loss = tf.reduce_mean(domains_loss)
+        self.loss = tf.reduce_sum(self.sentiment_loss + self.domains_loss)
```
Comment (Mostafa-Samir) [inline drnn.py]: Introduce `domain_loss_control` variable to stop training the domain NN after 10 epochs

In the beginning, define
```python
self.domain_loss_control = tf.placeholder(tf.int32, shape=(), name='domain_loss_control')
```
and then incorporate it in the loss as:
```python
self.loss = tf.reduce_sum(self.sentiment_loss + self.domain_loss_control * self.domain_loss)
```

Make the number of epochs to turn-off domain training (lets call it `k`) configurable and set it initially to 10 (or whatever value you think is better from your experience with previous training logs). The `self.domain_loss_control` will be fed with the value `1` for the first `k` epochs and then `0` for the rest of epochs

Comment (ibrahimsharaf) [reply drnn.py]: Makes sense, I guess setting `k` with 5 would suffice.
